vpr/src/place/analytical_placement/AnalyticalSolver.h

```diff
@@ -2,32 +2,49 @@
 #pragma once
 
 #include <memory>
-#include <vector>
-#include "PartialPlacement.h"
 #include "Eigen/Sparse"
+#include "ap_netlist_fwd.h"
+#include "vtr_strong_id.h"
+#include "vtr_vector.h"
+
+class PartialPlacement;
+class APNetlist;
 
 enum class e_analytical_solver {
     QP_HYBRID,
     B2B
 };
 
+// A new strong ID for the AP rows.
+// This gives a linearized ID for each of the moveable blocks from 0 to the
+// number of moveable blocks.
+struct ap_row_id_tag {};
+typedef vtr::StrongId<ap_row_id_tag, size_t> APRowId;
+
 // TODO: How should the hint be handled?
 //       Perhaps the PartialPlacement could be used as a hint. Could have a flag
 //       within detailing if the placement has been populated.
 class AnalyticalSolver {
 public:
     virtual ~AnalyticalSolver() {}
+    AnalyticalSolver(const APNetlist &inetlist);
     virtual void solve(unsigned iteration, PartialPlacement &p_placement) = 0;
+
+protected:
+    const APNetlist& netlist;
+    size_t num_moveable_blocks = 0;
```

Comment:
Should we have this field here, or should it be part of the netlist or the partialplacement?

Reply:
I have thought about this. The reason I brought this to the AnalyticalSolver class is that it is the only user of the Netlist that cares about the number of moveable blocks. Legalizers only care if a given block is moveable or not. I am trying to make the PartialPlacement simpler such that it ONLY contains the locations of the blocks at a given moment in time; nothing more. I feel like it could be easily added to the Netlist (in which the netlist would maintain it); but I just feel that it is not necessary.

In the future, if there is a need outside of the solver for this information, it can be moved into the APNetlist class. Very good point.